changes/views.py

```diff
@@ -46,3 +52,147 @@ class ChangeGetViewSet(viewsets.ReadOnlyModelViewSet):
     queryset = Change.objects.all()
     serializer_class = ChangeSerializer
     filter_class = ChangeFilter
+
+
+class ReceiveIdentityStoreOptout(mixins.CreateModelMixin,
+                                 generics.GenericAPIView):
+    permission_classes = (IsAuthenticated,)
+
+    def post(self, request, *args, **kwargs):
+        """Handles optout notifications from the Identity Store."""
+        try:
+            data = utils.json_decode(request.body)
+        except ValueError as e:
+            return JsonResponse({'reason': "JSON decode error",
+                                'details': six.text_type(e)}, status=400)
+
+        try:
+            identity_id = data['identity']
+            optout_reason = data['optout_reason']
+            optout_source = data['optout_source']
+        except KeyError as e:
+            return JsonResponse({'reason': '"identity", "optout_type", '
+                                 '"optout_reason" and "optout_source" must be '
+                                 'specified.'
+                                 }, status=400)
+
+        registration = Registration.objects.get(mother_id=identity_id)
+        if registration.data.get('msg_receiver'):
+            fire_optout_receiver_type_metric(registration.data['msg_receiver'])
+
+        fire_optout_reason_metric(optout_reason)
+
+        if registration.data.get('msg_type'):
+            fire_optout_message_type_metric(registration.data['msg_type'])
+
+        fire_optout_source_metric(optout_source)
+
+        return JsonResponse({})
+
+
+def fire_optout_reason_metric(reason):
+    from registrations.tasks import fire_metric
+
+    if reason not in settings.OPTOUT_REASONS:
+        reason = 'other'
+
+    fire_metric.apply_async(kwargs={
+        "metric_name": 'optout.reason.%s.sum' % reason,
+        "metric_value": 1.0
+    })
+
+    def search_optouts_reason():
+        result = utils.search_optouts({"reason": reason})
+        return len(list(result))
```

Comment:
This will cause us to load the entire list of results in memory before counting it. Maybe we can do something like `sum(1 for r in result)` to count it without loading it all in memory?